### DISCUSSION TITLE
PIP5048MS Energy distribution no entities in the drop down menu.

### DISCUSSION THREAD

**Discussion by CrAzYDr1veR** [syssi/esphome-pipsolar#81] [Q&A]
Hi guys,

In home assistant the pipsolar data shows fine (volts amps watts etc)  but there is no options on the energy dashboard drop down menu to add solar or battery.
After adding my shelly i can select in the drop down options the grid input.

One option shows:

> Expected data source not listed
> You’re configuring a statistic but you couldn’t find your source in the dropdown? That’s caused by a bug in the integration providing the entity. Integrations need to configure their entities correctly so Home Assistant knows that we need to track statistics for it and how.
> 
> Open an issue with the author of the integration and link them to https://developers.home-assistant.io/docs/core/entity/sensor#long-term-statistics.

How can i add the missing entities?

**Comment by CrAzYDr1veR** [comment]
![energy flow card](https://github.com/syssi/esphome-pipsolar/assets/47764335/50b63244-40ea-4cc7-bd0b-f03fb5d18b18)
nedded to add filters but still no way to see the energy in or out the battery

    pv_charging_power:
      id: inverter0_pv_charging_power
      name : "PV Charging Power"
      filters:
        - multiply: 0.001
      unit_of_measurement: kWh
      device_class: energy
      state_class: total_increasing
      accuracy_decimals: 3
     

I have beed told to make utility helpers or sensors...

**Comment by magnetus26** [comment]
i think you might need a smart shunt to measure the power in / out the battery. idk that the pipaolar has those sensor, remember they are based on your battery bank total kWh or Ah to be able to calculate hoe much in/out they have

**Reply by CrAzYDr1veR** [reply]
I thought you used only data from the pipsolar on your system.
there is a battery charge and battery discharge current, i just have to learn how to adapt all of it, or forget about it and use the tesla style energy card addon to see if it's better to work on.

This is burning my brain cells :P

Values are wrong, just testing the entries.
![energy4](https://github.com/syssi/esphome-pipsolar/assets/47764335/76f2dc2e-444c-43a5-8f80-23206e3b2080)

**Reply by magnetus26** [reply]
im not talking about the energy dashboard.
i mean in your pipsolar ivnerter yaml code.
there should be sensors or templates to measure your battery capacity

**Reply by CrAzYDr1veR** [reply]
there isn't just like the pv power it's just instant values and you put  "state_class: total_increasing" to have a total value of kwh,
the same could be done for the battery after converting p = A * V

**Reply by magnetus26** [reply]
i dont think that will work `id: inverter0_pv_charging_power` only works during the charge cycle and the value is not constant, it will change based upon a lot of conditions, like the amount of solar enegy, it could be really sunny and you value will be high, it culd be oovercast and the production will be less, if its rainy there will might be no production all all, and at night that value will be 0 so you kWh will be 0 while panels are not producing.

in my case i had to use a smart shunt, and created a couple of templates, one for `battery in` and another for `battery out`
but I still have to know the total Ah capacity of my battery bank. ( in my case is just 264Ah ).

maybe theres another solution but im really green at coding and it is my current solution for `Home Battery Storage` parameters

here is part of my code maybe you can do something with it. just remember that the `264` is my particilar case and that the sensors that im using are from another ESPHome entity:

`battery voltage sensor`
```
sensor:
  - platform: victron_smart_shunt
     battery_voltage:
      name: "Battery Voltage"
      id: bv
      accuracy_decimals: ${accuracy}
```

`consumed amp hous`
  ```
consumed_amp_hours:
      name: "consumed amp hours"
      id: consumed_amp_hours
      unit_of_measurement: Ah
      accuracy_decimals: ${accuracy}
```

I use this template as my Energy going in to battery or this will create adevice called  `SmartShunt charge` 
`battery in`
```
- platform: template
    name: " ${devicename} charge "
    filters:
      - multiply: 0.001
    unit_of_measurement: kWh
    device_class: energy
    state_class: total_increasing
    accuracy_decimals: ${accuracy}
    lambda: |-
          float i = 264 + id(consumed_amp_hours).state; 
          float e = id(bv).state;
          float w = 0.0;
          if (i > 0) {
            w =  i * e;
          }
          return {w} ;
```


I use this template as my Energy going in to battery or this will create adevice called  ` SmartShunt discharge `  
`battery out`
 ```
- platform: template
    name: " ${devicename} discharge "
    filters:
      - multiply: 0.001
    unit_of_measurement: kWh
    device_class: energy
    state_class: total_increasing
    accuracy_decimals: ${accuracy}
    lambda: |-
          float i = id(consumed_amp_hours).state;
          float e = id(bv).state;
          float w = 0.0;
          if (i < 0) {
            w = i * e;
          }
          return {w * -1} ;
```
